[AWS] [API Gateway] Add custom access logging fields (#7151)
elastic/integrations: packages/aws/_dev/build/docs/apigateway.md

**Logs** help you keep a record of events happening in Amazon API Gateway.
Logs collected by the Amazon API Gateway integration include the HTTP method, status code, request details and more. 

AWS API Gateway allows you to set up Custom Access Logging to gain deeper insights into the requests and responses through your API. (ex. [REST](https://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-mapping-template-reference.html#context-variable-reference-access-logging-only))

In addition to the default log format, you can enhance your log entries by leveraging the following context variables:

> ```{"apiId": "$context.apiId", "domainName": "$context.domainName", "stage": "$context.stage"}```

Reviewer: What about adding a JSON with ALL the fields supported by this integration, ready to be copied and pasted?

IIRC, it is essential to have a "core" set of fields that users can extend with additional ones, but users must always include the core fields, right? It can be helpful for the actual testing and as documentation of what's possible.

Can we include some sample documents with apiId and the other additional fields in the tests?

Author: 1. By enabling Custom Access Logging, users can select only the fields necessary, so it's not required to always have the core fields.

2. By not enabling Custom Access Logging users will fetch the default log format which is as follows per API type:
HTTP:

>   "requestId": "$context.requestId",
>   "ip": "$context.identity.sourceIp",
>   "requestTime": "$context.requestTime",
>   "httpMethod": "$context.httpMethod",
>   "routeKey": "$context.routeKey",
>   "status": "$context.status",
>   "protocol": "$context.protocol",
>   "responseLength": "$context.responseLength"

REST:

>   "requestId": "$context.requestId",
>   "ip": "$context.identity.sourceIp",
>   "caller": "$context.identity.caller",
>   "user": "$context.identity.user",
>   "requestTime": "$context.requestTime",
>   "httpMethod": "$context.httpMethod",
>   "resourcePath": "$context.resourcePath",
>   "status": "$context.status",
>   "protocol": "$context.protocol",
>   "responseLength": "$context.responseLength"

WebSocket:

>   "requestId": "$context.requestId",
>   "ip": "$context.identity.sourceIp",
>   "caller": "$context.identity.caller",
>   "user": "$context.identity.user",
>   "requestTime": "$context.requestTime",
>   "eventType": "$context.eventType",
>   "routeKey": "$context.routeKey",
>   "status": "$context.status",
>   "connectionId": "$context.connectionId"

Since some of the fields are only supported by specific API types, I could include the default log format for each in the documentation so there's no confusion. WDYT @zmoog?

Users are presented with the default log format by default when enabling Custom Access Logging and they can also reset back to it by pressing Insert Example, ex for REST:

<img width="464" alt="REST" src="https://github.com/elastic/integrations/assets/59661554/04dee67a-bf89-4d3d-99c7-6765cbb2bd9d">

<br></br>
> Can we include some sample documents with apiId and the other additional fields in the tests?

They are at the bottom of the logs since I also also kept the original tests 😄

Reviewer: Let's test my understanding.

When I visit: **REST API > Stages > {stage} > Logs/Tracing**, I can enable the access logs by selecting "Enable Access Logging". Other API types have a similar option.

By default, the "Log Format" is empty; the AWS web console can drop a JSON sample log format by clicking the "JSON" button.

As an alternative, we can provide a similar sample log format with all the fields for each type we currently support in our integration.

> Since some of the fields are only supported by specific API types, I could include the default log format for each in the documentation so there's no confusion.

Since different API types have various fields, having a recommended JSON log format for each API type would be great.

Reviewer: Got it. Thanks for clarifying!

If you decide not to use the default and provided naming conventions for logs,

> creating custom pipelines is required.

Reviewer: We should expand this section.

Here's the information we should convey to the users:

- the API Gateway offers a wide array of additional fields you can pick from
- you can leverage the custom pipeline to process the field names or values
- you can leverage the custom mapping to store data efficiently

REST:  
```{"requestId": "$context.requestId","ip": "$context.identity.sourceIp","caller": "$context.identity.caller","user": "$context.identity.user""requestTime": "$context.requestTime","httpMethod": "$context.httpMethod","resourcePath": "$context.resourcePath","status": "$context.status","protocol": "$context.protocol","responseLength": "$context.responseLength",apiId": "$context.apiId", "domainName": "$context.domainName", "stage": "$context.stage"}```

WebSockets:  
```{"requestId": "$context.requestId","ip": "$context.identity.sourceIp","caller": "$context.identity.caller","user": "$context.identity.user","requestTime": "$context.requestTime","eventType": "$context.eventType","routeKey": "$context.routeKey","status": "$context.status","connectionId": "$context.connectionId", apiId": "$context.apiId", "domainName": "$context.domainName", "stage": "$context.stage"}```

> To leverage the extensive range of additional fields offered by enabling Custom Access Logging, create custom pipelines for processing field names or values, alongside custom mapping to store data efficiently.

Author: good point. how about now @zmoog?

REST:  
```{"requestId": "$context.requestId","ip": "$context.identity.sourceIp","caller": "$context.identity.caller","user": "$context.identity.user""requestTime": "$context.requestTime","httpMethod": "$context.httpMethod","resourcePath": "$context.resourcePath","status": "$context.status","protocol": "$context.protocol","responseLength": "$context.responseLength",apiId": "$context.apiId", "domainName": "$context.domainName", "stage": "$context.stage"}```

WebSockets:  
```{"requestId": "$context.requestId","ip": "$context.identity.sourceIp","caller": "$context.identity.caller","user": "$context.identity.user","requestTime": "$context.requestTime","eventType": "$context.eventType","routeKey": "$context.routeKey","status": "$context.status","connectionId": "$context.connectionId", apiId": "$context.apiId", "domainName": "$context.domainName", "stage": "$context.stage"}```

> Beyond the standard Log Formats we provide, you have the option to use a wide range of fields for each API type when enabling Custom Access Logging.
> 
> To utilize these fields in the integration, create custom pipelines. These pipelines will process field names or values. You'll also need to create custom mapping to store your data efficiently. For more information on how to do this, check out our [official documentation](https://www.elastic.co/guide/en/fleet/current/data-streams-pipeline-tutorial.html).

Reviewer: ```suggestion
Beyond the standard Log Formats we provide, you can use a wide range of fields for each API type when enabling Custom Access Logging.

To utilize these fields in the integration, create custom pipelines. These pipelines will process field names or values. You'll also need to create a custom mapping to store your data efficiently. For more information on how to do this, check out our [official documentation](https://www.elastic.co/guide/en/fleet/current/data-streams-pipeline-tutorial.html).
```

REST:  
```{"requestId": "$context.requestId","ip": "$context.identity.sourceIp","caller": "$context.identity.caller","user": "$context.identity.user""requestTime": "$context.requestTime","httpMethod": "$context.httpMethod","resourcePath": "$context.resourcePath","status": "$context.status","protocol": "$context.protocol","responseLength": "$context.responseLength"}```

WebSockets:  
```{"requestId": "$context.requestId","ip": "$context.identity.sourceIp","caller": "$context.identity.caller","user": "$context.identity.user","requestTime": "$context.requestTime","eventType": "$context.eventType","routeKey": "$context.routeKey","status": "$context.status","connectionId": "$context.connectionId"}```

In addition to the Default Log Format, the integration supports the following context variables for all API types:  
```{"apiId": "$context.apiId", "domainName": "$context.domainName", "stage": "$context.stage"}```

> 

Reviewer: What about providing an "Elastic Log Format" instead of the default one, based on the default one + the extra variables (`appId`, `domainName`, and `stage`) for each API type to copy and paste?
